packages/medusa/src/services/cart.ts

```diff
@@ -1004,109 +1038,126 @@ class CartService extends BaseService {
    * @return the result of the update operation
    */
   async applyDiscount(cart: Cart, discountCode: string): Promise<void> {
-    const discount = await this.discountService_.retrieveByCode(discountCode, [
-      "rule",
-      "regions",
-    ])
-
-    if (cart.customer_id) {
-      const canApply = await this.discountService_.canApplyForCustomer(
-        discount.rule.id,
-        cart.customer_id
+    return this.atomicPhase_(async (manager) => {
+      const discount = await this.discountService_.retrieveByCode(
+        discountCode,
+        ["rule", "regions"]
       )
 
-      if (!canApply) {
-        throw new MedusaError(
-          MedusaError.Types.NOT_ALLOWED,
-          "Discount is not valid for customer"
+      if (cart.customer_id) {
+        const canApply = await this.discountService_.canApplyForCustomer(
+          discount.rule.id,
+          cart.customer_id
         )
+
+        if (!canApply) {
+          throw new MedusaError(
+            MedusaError.Types.NOT_ALLOWED,
+            "Discount is not valid for customer"
+          )
+        }
       }
-    }
 
-    const rule = discount.rule
+      const rule = discount.rule
 
-    // if limit is set and reached, we make an early exit
-    if (discount.usage_limit) {
-      discount.usage_count = discount.usage_count || 0
+      // if limit is set and reached, we make an early exit
+      if (discount.usage_limit) {
+        discount.usage_count = discount.usage_count || 0
 
-      if (discount.usage_count >= discount.usage_limit) {
+        if (discount.usage_count >= discount.usage_limit) {
+          throw new MedusaError(
+            MedusaError.Types.NOT_ALLOWED,
+            "Discount has been used maximum allowed times"
+          )
+        }
```

Comment:
nit: we should consider moving this logic to the discount service; maybe we should create a `canApply` method that handles everything from usage conditions to customer permissions to potentially "custom permissions" in the future. 

Don't expect you to include in this PR but noting it here so we can make a ticket for it

Reply:
+ date ranges

Reply:
By date ranges I mean - the `canApply` method should also take into account if the discount has expired/is scheduled. Sorry for the sub-par explanation 😅

Reply:
yo @srindom! did some changes, lmk what you think!